pkg/deploy/controller/deploymentconfig/controller_test.go

```diff
@@ -1,835 +1,629 @@
+				{version: 4, replicas: 1, replicasA: newint(1), status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 5, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: false},
+			},
+			errExpected: false,
+		},
+		{
+			name:             "(compat) already deployed, no active deployment",
+			replicas:         2,
+			newVersion:       2,
+			expectedReplicas: 1,
+			before: []deployment{
+				{version: 1, replicas: 0, desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: false},
+				{version: 2, replicas: 0, desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: false},
+			},
+			after: []deployment{
+				{version: 1, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: false},
+				{version: 2, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: false},
+			},
+			errExpected: false,
+		},
+		{
+			name:             "(compat) scale up latest/active completed deployment",
+			replicas:         1,
+			newVersion:       2,
+			expectedReplicas: 5,
+			before: []deployment{
+				{version: 1, replicas: 0, status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 5, status: deployapi.DeploymentStatusComplete, cancelled: false},
+			},
+			after: []deployment{
+				{version: 1, replicas: 0, replicasA: newint(0), status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 5, replicasA: newint(5), status: deployapi.DeploymentStatusComplete, cancelled: false},
+			},
+			errExpected: false,
+		},
+		{
+			name:             "(compat) scale up active (not latest) completed deployment (RC targetted directly)",
+			replicas:         1,
+			newVersion:       2,
+			expectedReplicas: 5,
+			before: []deployment{
+				{version: 1, replicas: 5, status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 0, desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
+			},
+			after: []deployment{
+				{version: 1, replicas: 5, replicasA: newint(5), status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
+			},
+			errExpected: false,
+		},
+		/*
+			// TODO: In this case, I'm not sure how we can detect that the client has
+			// manually scaled the latest RC directly: the replica count might be less
+			// than desired if the deployment partially succeeded during a failure or
+			// cancellation, and it might be greater than desired due to a strategy
+			// bug (although the latter case seems like a more likely candidate for an
+			// assumption, which could maybe fix scale up most of the time but still
+			// leaves scale down broken.)
```

Comment:
That's correct: we don't know whether to adjust the RC to match the DC or
vice versa.

On Tue, Dec 1, 2015 at 8:56 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/deploy/controller/deploymentconfig/controller_test.go
> <https://github.com/openshift/origin/pull/5875#discussion_r46279924>:
>
> > +				{version: 2, replicas: 0, desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
> > +			},
> > +			after: []deployment{
> > +				{version: 1, replicas: 5, replicasA: newint(5), status: deployapi.DeploymentStatusComplete, cancelled: false},
> > +				{version: 2, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
> > +			},
> > +			errExpected: false,
> > +		},
> > +		/*
> > +			// TODO: In this case, I'm not sure how we can detect that the client has
> > +			// manually scaled the latest RC directly: the replica count might be less
> > +			// than desired if the deployment partially succeeded during a failure or
> > +			// cancellation, and it might be greater than desired due to a strategy
> > +			// bug (although the latter case seems like a more likely candidate for an
> > +			// assumption, which could maybe fix scale up most of the time but still
> > +			// leaves scale down broken.)
>
> So we can't impose the invariant that the active deployment is always
> converged to the DC (because the annotation on it still remains)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5875/files#r46279924>.
>